Frameworks/CoreGraphics/CGGraphicBufferImage.mm

```diff
@@ -30,9 +40,140 @@
+    HRESULT STDMETHODCALLTYPE GetStride(_Out_ UINT* stride);
+    HRESULT STDMETHODCALLTYPE GetDataPointer(_Out_ UINT* bufferSize, _Outptr_ WICInProcPointer* data);
+    HRESULT STDMETHODCALLTYPE GetPixelFormat(_Out_ WICPixelFormatGUID* pixelFormat);
+
+private:
+    CGGraphicBufferWICBitmap* m_imageWICBitmap;
+};
+
+class CGGraphicBufferWICBitmap : public RuntimeClass<Microsoft::WRL::RuntimeClassFlags<Microsoft::WRL::RuntimeClassType::ClassicCom>,
+                                                     IAgileObject,
+                                                     Microsoft::WRL::FtmBase,
+                                                     IWICBitmap> {
+public:
+    CGGraphicBufferWICBitmap();
+
+    HRESULT RuntimeClassInitialize(
+        _In_ UINT width, _In_ UINT height, _In_ UINT stride, _In_ WICPixelFormatGUID pixelFormat, _In_ BYTE* dataBuffer);
+
+    // IWICBitmap interface
+    HRESULT STDMETHODCALLTYPE Lock(_In_ const WICRect* lockRect, _In_ DWORD flags, _COM_Outptr_ IWICBitmapLock** outLock) {
+        ComPtr<IWICBitmapLock> lock;
+        MakeAndInitialize<CGGraphicBufferWICBitmapLock>(&lock, this);
+        *outLock = lock.Detach();
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE SetPalette(_In_ IWICPalette* palette) {
+        return E_NOTIMPL;
+    }
+
+    HRESULT STDMETHODCALLTYPE SetResolution(_In_ double dpiX, _In_ double dpiY) {
+        return E_NOTIMPL;
+    }
+
+    // IWICBitmapSource interface (that IWICBitmap inherits)
+    HRESULT STDMETHODCALLTYPE GetSize(_Out_ UINT* width, _Out_ UINT* height) {
+        *width = m_width;
+        *height = m_height;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetPixelFormat(_Out_ WICPixelFormatGUID* pixelFormat) {
+        *pixelFormat = m_pixelFormat;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE GetResolution(_Out_ double* dpiX, _Out_ double* dpiY) {
+        return E_NOTIMPL;
+    }
+
+    HRESULT STDMETHODCALLTYPE CopyPalette(_In_ IWICPalette* palette) {
+        return E_NOTIMPL;
+    }
+
+    HRESULT STDMETHODCALLTYPE CopyPixels(_In_opt_ const WICRect* copyRect,
+                                         _In_ UINT stride,
+                                         _In_ UINT bufferSize,
+                                         _Out_writes_all_(cbBufferSize) BYTE* buffer) {
+        return E_NOTIMPL;
+    }
+
+public:
+    UINT m_width;
+    UINT m_height;
+    UINT m_stride;
+    WICPixelFormatGUID m_pixelFormat;
+    BYTE* m_dataBuffer;
+};
+
+CGGraphicBufferWICBitmap::CGGraphicBufferWICBitmap()
+    : m_width(0), m_height(0), m_stride(0), m_pixelFormat(GUID_WICPixelFormat32bppPBGRA), m_dataBuffer(nullptr) {
+}
+
+HRESULT CGGraphicBufferWICBitmap::RuntimeClassInitialize(
+    UINT width, UINT height, UINT stride, WICPixelFormatGUID pixelFormat, BYTE* dataBuffer) {
+    m_width = width;
+    m_height = height;
+    m_stride = stride;
+    m_pixelFormat = pixelFormat;
+    m_dataBuffer = dataBuffer;
+    return S_OK;
+}
+
```

Comment:
Ah yes I remember teaching me this long back

---
In reply to: [80783282](https://github.com/Microsoft/WinObjC/pull/1040#discussion_r80783282) [](ancestors = 80783282)